Framework/Kernel/src/TimeSplitter.cpp

```diff
@@ -0,0 +1,189 @@
+    if ((startIterator->first != start) && (startIterator != m_roi_map.begin()))
+      startIterator--; // move to the one before
+
+    // the end is one past the "stop"
+    auto stopIterator = m_roi_map.upper_bound(stop);
+    if ((stopIterator != m_roi_map.end()) && (stopValue == IGNORE_VALUE))
+      stopIterator++; // move to the one after
+
+    const bool atStart = (startIterator == m_roi_map.begin());
+
+    // remove the elements that are being replaced [inclusive, exclusive)
+    m_roi_map.erase(startIterator, stopIterator);
+
+    // put in the new elements
+    if ((value > IGNORE_VALUE) || (!atStart)) {
+      if (value != this->valueAtTime(start)) {
+        m_roi_map.insert({start, value});
+      }
+    }
+
+    // find the new iterator for where this goes to see if it the same value
+    stopIterator = m_roi_map.lower_bound(stop);
+    if ((stopIterator != m_roi_map.end()) && (value == stopIterator->second)) {
+      m_roi_map.erase(stopIterator);
+    }
+    if (value != stopValue) {
+      m_roi_map.insert({stop, stopValue});
+    }
+
+    // verify this ends with IGNORE_VALUE
+    if (m_roi_map.rbegin()->second != IGNORE_VALUE) {
+      throw std::runtime_error("Something went wrong in TimeSplitter::addROI");
+    }
+  }
+}
+
+void TimeSplitter::clearAndReplace(const Types::Core::DateAndTime &start, const Types::Core::DateAndTime &stop,
+                                   const int value) {
+  m_roi_map.clear();
+  if (value >= 0) {
+    m_roi_map.insert({start, value});
+    m_roi_map.insert({stop, IGNORE_VALUE});
+  }
+}
+
+int TimeSplitter::valueAtTime(const Types::Core::DateAndTime &time) const {
+  if (m_roi_map.empty())
+    return IGNORE_VALUE;
+  if (time < m_roi_map.begin()->first)
+    return IGNORE_VALUE;
+
+  // this method can be used when the object is in an unusual state and doesn't
+  // end with IGNORE_VALUE
+
+  // find location that is greater than or equal to the requested time and give
+  // back previous value
+  auto location = m_roi_map.lower_bound(time);
+  if (location->first == time) {
+    // found the time in the map
+    return location->second;
+  } else if (location == m_roi_map.begin()) {
+    // iterator is greater than the first value in the map b/c equal is already
+    // handled asked for a time outside of the map
+    return IGNORE_VALUE;
+  } else {
+    // go to the value before
+    location--;
+    return location->second;
+  }
+}
+
+/**
+ * Return a sorted vector of the output workspace indices
+ */
+std::vector<int> TimeSplitter::outputWorkspaceIndices() const {
+  // sets have unique values and are sorted
+  std::set<int> outputSet;
+
+  // copy all of the (not ignore) output workspace indices
+  for (const auto iter : m_roi_map) {
+    if (iter.second > IGNORE_VALUE)
+      outputSet.insert(iter.second);
+  }
+
+  // return a vector
+  return std::vector<int>(outputSet.begin(), outputSet.end());
+}
+
+/**
+ * Returns a Mantid::Kernel::TimeROI for the requested workspace index.
```

Comment:
Are the cases mentioned (workspaceindex < 0 or non-existent workspaceindex) an input error or valid inputs? If the former, the function should raise an exception instead of returning and empty ROI

Reply:
OK, but a destination index of `-1` _means_ a time span that should be ignored, so it does have a meaning and therefore the code should _not_ raise an exception.